### PR TITLE
feat: add docs for Dal

### PR DESCRIPTION
**Pull request by ChengYueJia**
## What ❔

add docs and diagrams for dal.

## Why ❔
Easier to understand data workflow managed in DAL.
Easier to debug.
Easier to maintain and add features for guys who wanna leverage the power of zkStack.

* https://github.com/matter-labs/zksync-era/issues/954

## Checklist

- [ ] PR title corresponds to the body of PR (we generate changelog entries from PRs).
- [ ] Tests for the changes have been added / updated.
- [ ] Documentation comments have been added / updated.
- [ ] Code has been formatted via `zk fmt` and `zk lint`.
- [ ] Spellcheck has been run via `zk spellcheck`.
- [ ] Linkcheck has been run via `zk linkcheck`.

### CODE REVIEW

File: core/lib/dal/doc/FriProofCompressorDal.md [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+stateDiagram-v2
+[*] --> queued : insert_proof_compression_job
+queued --> in_progress : get_next_proof_compression_job
+in_progress --> successful : mark_proof_compression_job_successful
+successful --> [*] 
+in_progress --> failed : mark_proof_compression_job_failed
+failed --> [*]
+failed --> queued : requeue_stuck_jobs
+
+successful --> sent_to_server : get_least_proven_block_number_not_sent_to_server
```
Comment (perekopskiy) [inline core/lib/dal/doc/FriProofCompressorDal.md:23]: `get_least_proven_block_number_not_sent_to_server` doesn't mutate state, it should be `mark_proof_sent_to_server` instead

File: core/lib/dal/doc/FriProofCompressorDal.md [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+successful --> skipped : mark_proof_compression_job_successful
```
Comment (perekopskiy) [inline core/lib/dal/doc/FriProofCompressorDal.md:25]: This edge is wrong

File: core/lib/dal/doc/FriProofCompressorDal.md [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+---
+title: Status Diagram
+---
+stateDiagram-v2
+[*] --> queued : insert_proof_compression_job
+queued --> in_progress : get_next_proof_compression_job
+in_progress --> successful : mark_proof_compression_job_successful
+successful --> [*] 
+in_progress --> failed : mark_proof_compression_job_failed
+failed --> [*]
```
Comment (perekopskiy) [inline core/lib/dal/doc/FriProofCompressorDal.md:20]: This edge is wrong

File: core/lib/dal/doc/FriProofCompressorDal.md [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+failed --> queued : requeue_stuck_jobs
+
+successful --> sent_to_server : get_least_proven_block_number_not_sent_to_server
+
+successful --> skipped : mark_proof_compression_job_successful
+sent_to_server --> [*] : mark_proof_sent_to_server
```
Comment (perekopskiy) [inline core/lib/dal/doc/FriProofCompressorDal.md:26]: This should be without `: mark_proof_sent_to_server`

File: core/lib/dal/doc/FriProofCompressorDal.md [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+
+[*] --> skipped : skip_proof_compression_job
+skipped --> [*]
```
Comment (perekopskiy) [inline core/lib/dal/doc/FriProofCompressorDal.md:29]: This edge is wrong

File: core/lib/dal/doc/FriProverDal.md [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+stateDiagram-v2
+[*] --> queued : insert_prover_job
+queued --> in_progress : get_next_job
+in_progress --> successful : save_proof
+successful --> [*] 
+in_progress --> failed : save_proof_error
+failed --> [*]
+failed --> queued : requeue_stuck_jobs
+
+successful --> sent_to_server : save_successful_sent_proof
```
Comment (perekopskiy) [inline core/lib/dal/doc/FriProverDal.md:23]: `sent_to_server` should be removed, it's not achiavable

**Review by perekopskiy — COMMENTED** [review]
Thanks for contribution, nice diagrams!

I've left some comments.
General comment: `requeue_*` methods can move not only failed -> queued but also in_progress -> queued. Please fix where applicable

File: core/lib/dal/doc/FriProverDal.md (hunk 6 above)
Comment (ChengYueJia) [reply core/lib/dal/doc/FriProverDal.md:24]: @perekopskiy Hi, I've updated some of feedbacks. Please have a view

**Comment by ChengYueJia** [comment]
@perekopskiy Hi, I've updated some of feedbacks. Please have a view

File: core/lib/dal/doc/README.md [hunk 7]
```diff
@@ -0,0 +1,2 @@
+#
```
Comment (perekopskiy) [inline core/lib/dal/doc/README.md:1]: Is this file needed?

**Review by perekopskiy — COMMENTED** [review]
Please run `zk fmt` to format added files

**Comment by ChengYueJia** [comment]
> Please run `zk fmt` to format added files

It's done

**Review by perekopskiy — APPROVED** [review]
